### PR TITLE
Add /v2/leader API endpoint which abdicates leadership on DELETE

### PR DESCRIPTION
**Pull request by sttts**
During testing of failover scenarios one has to force a leadership change.
This API endpoint allows to force the current leader to abdictate its
leadership and start a new leadership election.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class LeaderResource @Inject() (
+    schedulerService: MarathonSchedulerService,
+    val config: MarathonConf with HttpConf)
+    extends RestResource {
+
+  @DELETE
+  @Produces(Array(MediaType.APPLICATION_JSON))
+  def delete(): Response = {
+    schedulerService.isLeader match {
+      case false => notFound(s"Marathon instance '${config.hostname}' is no leader")
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala]: _is no leader_ => _is not the leader._

File: src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import mesosphere.chaos.http.HttpConf
+import mesosphere.marathon.api.RestResource
+import mesosphere.marathon.{ MarathonSchedulerService, MarathonConf }
+
+@Path("v2/leader")
+class LeaderResource @Inject() (
+    schedulerService: MarathonSchedulerService,
+    val config: MarathonConf with HttpConf)
+    extends RestResource {
+
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala]: Could we also add a handler for the GET verb while we're in here?

- If there's a leader, respond with `{ "leader": "..." }`
- If there's no leader, respond with a `404`

**Comment by ConnorDoyle** [comment]
LGTM. :ship: it after the comments are addressed.

File: src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala (hunk 1 above)
Comment (sttts) [inline src/main/scala/mesosphere/marathon/api/v2/LeaderResource.scala]: There is a contradiction also in the /v2/info resource: all requests are forwarded to the leader anyway. Hence, either the REST service responds, then it's the leader answering. Or the REST services doesn't respond, then there is no leader.

**Comment by rasputnik** [comment]
Due to the fact that these requests get forwarded to the leader,
won't it be a bit confusing to send a DELETE request to

http://not-master:8080/v2/leader

and have a different machine abdicate?

On 10 January 2015 at 10:37, Dr. Stefan Schimanski <notifications@github.com

**Comment by sttts** [comment]
But it's called "leader", not "leadership". So IMO the logic is fine.

**Comment by drexin** [comment]
The logic in the code looks like: "Only abdicate the leadership if you are the leader", but due to the fact that no non-leader will ever receive this call, in reality it is "delete the leader, whoever it is". Two possible solutions:

1. Create a special rule in the filter to process these requests on the machine that receives it
2. Rename it to "leader re-election" and expect it to always be processed by the leader

@ConnorDoyle WDYT?

**Comment by ConnorDoyle** [comment]
I'm OK with the current naming and behavior, provided the response is updated for the case where there is no leader.  Instead of advising that the current node is not the leader (which no longer makes sense in light of the proxy behavior), we should respond with the same 404 message as we do for `GET /v2/leader`.  We can also make the behavior a little more explicit in the API doc.

**Comment by sttts** [comment]
Updated the 404 strings and the made the documentation more verbose.

**Comment by drexin** [comment]
:ship:

**Pull request merged**